db/src/db.rs

```diff
@@ -899,135 +970,216 @@ mod tests {
-                let (report, next_partition_map, next_schema) = maybe_report.unwrap();
-                *partition_map = next_partition_map;
-                if let Some(next_schema) = next_schema {
-                    *schema = next_schema;
-                }
-
-                assert_eq!(index, report.tx_id,
-                           "\n{} - expected tx_id {} but got tx_id {}", label, index, report.tx_id);
-
-                let transactions = debug::transactions_after(&conn, &schema, index - 1).unwrap();
-                assert_eq!(transactions.0.len(), 1);
-                assert_eq!(*expected_transaction,
-                           transactions.0[0].into_edn(),
-                           "\n{} - expected transaction:\n{}\nbut got transaction:\n{}", label, *expected_transaction, transactions.0[0].into_edn());
-            }
-
-            if let Some(expected_datoms) = expected_datoms {
-                let datoms = debug::datoms_after(&conn, &schema, bootstrap::TX0).unwrap();
-                assert_eq!(*expected_datoms,
-                           datoms.into_edn(),
-                           "\n{} - expected datoms:\n{}\nbut got datoms:\n{}", label, *expected_datoms, datoms.into_edn())
-            }
-
-            // Don't allow empty tests.  This will need to change if we allow transacting schema
-            // fragments in a preamble, but for now it might catch malformed tests.
-            assert_ne!((expected_transaction, expected_datoms), (None, None),
-                       "Transaction test must include at least one of :test/expected-transaction or :test/expected-datoms");
-        }
-    }
-
     #[test]
     fn test_add() {
-        let mut conn = new_connection("").expect("Couldn't open in-memory db");
-        let mut db = ensure_current_version(&mut conn).unwrap();
-
-        // Does not include :db/txInstant.
-        let datoms = debug::datoms_after(&conn, &db.schema, 0).unwrap();
-        assert_eq!(datoms.0.len(), 88);
-
-        // Includes :db/txInstant.
-        let transactions = debug::transactions_after(&conn, &db.schema, 0).unwrap();
-        assert_eq!(transactions.0.len(), 1);
-        assert_eq!(transactions.0[0].0.len(), 89);
-
-        // TODO: extract a test macro simplifying this boilerplate yet further.
-        let value = edn::parse::value(include_str!("../../tx/fixtures/test_add.edn")).unwrap().without_spans();
-
-        let transactions = value.as_vector().unwrap();
-
-        assert_transactions(&conn, &mut db.partition_map, &mut db.schema, transactions);
+        let mut conn = TestConn::default();
+
+        // Test inserting :db.cardinality/one elements.
+        conn.transact("[[:db/add 100 :db/ident :keyword/value1]
+                        [:db/add 101 :db/ident :keyword/value2]]").unwrap();
+        assert_matches!(conn.last_transaction(),
+                        "[[100 :db/ident :keyword/value1 ?tx true]
+                          [101 :db/ident :keyword/value2 ?tx true]
```

Comment:
Yes, I think your understanding is correct.  The pattern above would match
```edn
[[100 :db/ident :keyword/value1 12345 true]
 [100 :db/ident :keyword/value2 12345 true]]
```
but not
```edn
[[100 :db/ident :keyword/value1 12345 true]
 [100 :db/ident :keyword/value2 99999 true]]
```